Mathlib/Topology/Kuratowski.lean

```diff
@@ -0,0 +1,127 @@
+/-
+Copyright (c) 2025 Jan Förster, Leon Müller, Luis Sand, and Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jan Förster, Leon Müller, Luis Sand, Junyan Xu
+-/
+import Mathlib.Data.Set.Card
+import Mathlib.Topology.Closure
+/-!
+# The Kuratowski closure-complement theorem
+
+This file proves Kuratowski's closure-complement theorem, which says that if one repeatedly
+applies the closure and complement operators to a set in a topological space, at most 14
+distinct sets can be obtained.
+
+In another file it will be shown that the maximum can be realized in the real numbers.
+(A set realizing the maximum is called a 14-set.)
+
+## Main definitions
+
+* `Topology.ClosureCompl.IsObtainable` is the binary inductive predicate saying that a set can be
+  obtained from another using the closure and complement operations.
+
+* `Topology.ClosureCompl.theFourteen` is an explicit multiset of fourteen sets obtained from a
+  given set using the closure and complement operations.
+
+## Main results
+
+* `Topology.ClosureCompl.mem_theFourteen_iff_isObtainable`:
+  a set `t` is obtainable by from `s` if and only if they are in the Multiset `theFourteen s`.
+
+* `Topology.ClosureCompl.ncard_isObtainable_le_fourteen`:
+  for an arbitrary set `s`, there are at most 14 distinct sets that can be obtained from `s` using
+  the closure and complement operations (the **Kuratowski closure-complement theorem**).
+
+## Notations
+
+* `k`: the closure of a set.
+* `i`: the interior of a set.
+* `c`: the complement of a set in docstrings; `ᶜ` is used in the code.
+
+## References
+
+* https://en.wikipedia.org/wiki/Kuratowski%27s_closure-complement_problem
+* https://web.archive.org/web/20220212062843/http://nzjm.math.auckland.ac.nz/images/6/63/The_Kuratowski_Closure-Complement_Theorem.pdf
+-/
+
+namespace Topology.ClosureCompl
+
+variable {X : Type*} [TopologicalSpace X]
+
+/-- `IsObtainable s t` means that `t` can be obtained from `s` by taking closures and complements.
+-/
+inductive IsObtainable (s : Set X) : Set X → Prop
+  | base : IsObtainable s s
+  | closure ⦃t : Set X⦄ : IsObtainable s t → IsObtainable s (closure t)
+  | complement ⦃t : Set X⦄ : IsObtainable s t → IsObtainable s tᶜ
+
+local notation "k" => closure
+local notation "i" => interior
+
+/-- The function `kckc` is idempotent: `kckckckc s = kckc s` for any set `s`.
+This is because `ckc = i` and `ki` is idempotent. -/
+theorem kckc_idem (s : Set X) : k (k (k (k sᶜ)ᶜ)ᶜ)ᶜ = k (k sᶜ)ᶜ := by
+  simp_rw [← interior_eq_compl_closure_compl, closure_interior_idem]
```

Comment:
I have to say, I don't love that we're naming things based on the notation. I considered making `k` and `c` (or the combination `kc`) abbreviations instead, but decided it was probably worse.
```suggestion
theorem kckc_idem (s : Set X) : k (k (k (k sᶜ)ᶜ)ᶜ)ᶜ = k (k sᶜ)ᶜ := by simp
```